### PR TITLE
Gallery: Move "Add new" action to Toolbar

### LINKED ISSUE #20837: Cover: Move color controls to the toolbar
Color controls, like background media, are primary features of this block and should be moved to the toolbar:

image

### LINKED ISSUE #21247: Move Gallery "Add new / Media Library" actions to the Toolbar
Right now we have a trailing appender on the Gallery block which is a bit problematic — it only allows inserting new images or editing the gallery if you are at the bottom of it. On large galleries that requires scrolling. We should move these controls to the toolbar instead, which scrolls with the user.

### PR DESCRIPTION
**Pull request by obenland**
## Description

Adds a block control to add/upload images to existing galleries.
Removes the dropzone appender at the end of galleries.

Fixes #21247.

## Screenshots 

Before | After
--- | ---
Screen Shot 2020-09-14 at 4 10 18 PM | Screen Shot 2020-09-14 at 4 10 54 PM

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by obenland** [comment]
Next steps:
* Update CSS classes
* Separate out dropdown control into its own component

**Comment by obenland** [comment]
Not sure if the suggested way to do this is the best, any feedback you have I'd love to hear.
Another area of uncertainty is naming 🙂
`MediaAddButton` sounds a bit awkward but feels close to the existing `media-*` nomenclature.

**Review by cpapazoglou — APPROVED** [review]
Tested by test instructions and LGTM!

- [x] Removes the trailing appender
- [x] Introduces an "Add" button with a text label 
- [x] "Add" button invokes a dropdown menu with "Open Media Library" and "Upload" options

Naming wise, I would argue that `media-add-flow` and `MediaAddFlow` would follow the pattern of `media-replace-flow` and `MediaReplaceFlow`. In that sense, it may also best fit under `packages/block-editor/src/components/media-add-flow/` dir serving as a component which may be reused from other media galleries perhaps (in the future)?

Should we consider exploring the fact that `MediaReplaceFlow` may be reused with some minor changes? It already accepts a Name attribute which can be set to `Add` and it seems we just need to return an array of media instead of a single media item. Apart from that, they seem functioning almost the same (I may be wrong of course). In that case, it could be also renamed to `MediaFlow`.

```
<MediaReplaceFlow
	mediaId={ ids }
	allowedTypes={ ALLOWED_MEDIA_TYPES }
	accept="image/*"
	name={ __( 'Add' ) }
	onSelect={ this.onSelectImages }
	onError={ this.onUploadError }
/>
```

**Comment by obenland** [comment]
I wonder if `MediaReplaceFlow` is a little too opinionated to be extendable like that. Putting `MediaAdd` alongside it might be a better option

File: packages/block-library/src/gallery/media-add.js [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+	ToolbarGroup,
+} from '@wordpress/components';
+import { useSelect } from '@wordpress/data';
+import { createRef } from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
+import { media as mediaIcon, upload } from '@wordpress/icons';
+import { DOWN } from '@wordpress/keycodes';
+
+export default function MediaAddButton( props ) {
+	const { allowedTypes, images, onError, onSelect, value } = props;
```
Comment (ZebulanStanphill) [inline packages/block-library/src/gallery/media-add.js:20]:
```suggestion
export default function MediaAddButton( {
	allowedTypes,
	images,
	onError,
	onSelect,
	value
} ) {
```

**Comment by cpapazoglou** [comment]
> I wonder if `MediaReplaceFlow` is a little too opinionated to be extendable like that. Putting `MediaAdd` alongside it might be a better option

Can't find an evidence of being opinionated, it seems that both add and replace do the same exact thing with different labelling. Not 100% sure though!

**Review by afercia — CHANGES_REQUESTED** [review]
From an a11y perspective:

1. The "Add" button should have an `aria-haspopup="true"` to announce it opens an additional UI, like the "More options" button on its right.  See also #24796 for the general issue.
2. Since there's a lot of space in the toolbar, can we make the "Add" text more explicit to better differentiate this button from all the other "Add block", "Add this", "Add that" ones?

Screenshot 2020-09-18 at 15 02 20

Changing the text tp "Add image" would be great.

General:
1. When opening the media modal, the dropdown menu stays on top of the media modal content (higher z-index I guess), at least for me. Screenshot:

Screenshot 2020-09-18 at 15 01 58


2. I see the `input type="file"` is hidden with `display: none;`. Though this relates to the specific component, are we sure it works correctly with all browsers? Recently there has been a similar issue in core where with old Edge and iOS the upload button didn't work. Also, this seems inconsistent with what core does. See https://core.trac.wordpress.org/changeset/47834 and https://core.trac.wordpress.org/ticket/49753 Cc @tellthemachines and @azaozz who know more about this issue.

**Comment by obenland** [comment]
@afercia Thank you for this excellent review and feedback!

I added the a11y improvements you pointed out (I was not aware of `aria-haspopup`, TIL!).

The dropdown menu staying on top of the media model is a mystery to me. I have no idea how I missed that during initial development, and I'm not sure how to best go about that. It looks like the z-index values for both components haven't changed in years, so I wonder if there's something that changed with toggling that dropdown open and close?
@ellatrix Do you happen to have any advice for me on how to fix that?

In regards to the `input type="file"` being hidden, yes, [that style is being set within the component](https://github.com/WordPress/gutenberg/blob/3e52a4da9e1f5d86a6084ad12192cee13c0721d6/packages/components/src/form-file-upload/index.js#L38). Since this PR doesn't really change anything in `FormFileUpload`, would you mind if we kept that separate for now?

**Comment by glendaviesnz** [comment]
@obenland with the [merging of the gallery refactor](https://github.com/WordPress/gutenberg/pull/25940), the `packages/block-library/src/gallery/edit.js` has radically changed, so it will be easier to reapply your changes onto the new file instead of trying to resolve the conflicts. 

Do you want to do that on this PR, or will it be easier to close this and create a new PR for it?

**Comment by glendaviesnz** [comment]
Going to close this for now. The [original issue](https://github.com/WordPress/gutenberg/issues/21247) has been added the the [gallery refactor project board](https://github.com/WordPress/gutenberg/projects/56) and we will reimplement this in a new PR against the refactored gallery block.

**Pull request closed without merging**